zerver/data_import/mattermost.py

```diff
@@ -0,0 +1,540 @@
+                user["is_mirror_dummy"] = True
+
+def reset_mirror_dummy_users(username_to_user: Dict[str, Dict[str, Any]]) -> None:
+    for username in username_to_user:
+        user = username_to_user[username]
+        user["is_mirror_dummy"] = False
+
+def do_convert_data(mattermost_data_file: str, output_dir: str, masking_content: bool) -> None:
+    mattermost_data = {}  # type: Dict[str, List[Dict[str, Any]]]
+    username_to_user = {}  # type: Dict[str, Dict[str, Any]]
+
+    os.makedirs(output_dir, exist_ok=True)
+    if os.listdir(output_dir):
+        raise Exception("Output directory should be empty!")
+
+    mattermost_data["version"] = []
+    mattermost_data["team"] = []
+    mattermost_data["channel"] = []
+    mattermost_data["user"] = []
+    mattermost_data["post"] = []
+
+    with open(mattermost_data_file, "r") as fp:
+        for line in fp:
+            row = ujson.loads(line.rstrip("\n"))
+            data_type = row["type"]
+            mattermost_data[data_type].append(row[data_type])
+
+    initialize_username_to_user_mapping(mattermost_data["user"], username_to_user)
+
+    realm_id = 0
+    for team in mattermost_data["team"]:
+        team_name = team["name"]
+
+        user_handler = UserHandler()
+        subscriber_handler = SubscriberHandler()
+        user_id_mapper = IdMapper()
+        stream_id_mapper = IdMapper()
+
+        print("Generating data for", team_name)
+        realm = make_realm(realm_id, team)
+        realm_output_dir = os.path.join(output_dir, team_name)
+
+        reset_mirror_dummy_users(username_to_user)
+        label_mirror_dummy_users(team_name, mattermost_data, username_to_user)
+
+        convert_user_data(
+            user_handler=user_handler,
+            user_id_mapper=user_id_mapper,
+            user_data_map=username_to_user,
+            realm_id=realm_id,
+            team_name=team_name,
+        )
+
+        zerver_stream = convert_channel_data(
+            channel_data=mattermost_data["channel"],
+            user_data_map=username_to_user,
+            subscriber_handler=subscriber_handler,
+            stream_id_mapper=stream_id_mapper,
+            user_id_mapper=user_id_mapper,
+            realm_id=realm_id,
+            team_name=team_name,
+        )
+
+        realm['zerver_stream'] = zerver_stream
+
+        all_users = user_handler.get_all_users()
+
+        zerver_recipient = build_recipients(
+            zerver_userprofile=all_users,
+            zerver_stream=zerver_stream,
+        )
+        realm['zerver_recipient'] = zerver_recipient
+
+        stream_subscriptions = build_stream_subscriptions(
+            get_users=subscriber_handler.get_users,
+            zerver_recipient=zerver_recipient,
+            zerver_stream=zerver_stream,
+        )
+
+        personal_subscriptions = build_personal_subscriptions(
+            zerver_recipient=zerver_recipient,
+        )
+
+        zerver_subscription = personal_subscriptions + stream_subscriptions
```

Comment:
Does Mattermost have a "group private message" feature?  I don't see code for importing data from such a thing here.  (HipChat doesn't have one, and Slack doesn't export PMs at all, so we don't have direct prior art for this).

Reply:
Mattermost support only exporting messages from channels right now.

> Posts (Posts in the Public/Private channels and also replies to those posts)

https://docs.mattermost.com/administration/bulk-export.html#bulk-export-tool

Reply:
OK, great, let's just mention that in the /help/ caveats list as something Mattermost doesn't export.

Reply:
(And maybe leave a comment here as well that this is where we'd include huddles if mattermost exported them)